File: android/src/main/kotlin/com/faizan/flutter_screen_time/AndroidOverlayConfiguration.kt
package com.faizan.flutter_screen_time

/**
 * Typed representation of the payload sent by `configureAndroidOverlayUI` from
 * Dart. Every field is nullable: an absent value means "keep the default", which
 * is applied later when the overlay is read back via
 * [ScreenTimePrefs.getShieldConfiguration].
 *
 * Mirrors the Dart `AndroidOverlayConfiguration` class.
 */
internal data class AndroidOverlayConfiguration(
    val title: String?,
    val message: String?,
    val backgroundColorHex: String?,
    val titleColorHex: String?,
    val messageColorHex: String?,
    val primaryButtonLabel: String?,
    val primaryButtonBackgroundColorHex: String?,
    val primaryButtonTextColorHex: String?,
    val secondaryButtonLabel: String?,
    val secondaryButtonBackgroundColorHex: String?,
    val secondaryButtonTextColorHex: String?,
    val titleFontSize: Float?,
    val messageFontSize: Float?,
    val cornerRadius: Float?,
) {
    companion object {
        /** Builds a configuration from the raw method-channel argument map. */
        fun fromMap(map: Map<String, Any?>): AndroidOverlayConfiguration {
            fun string(key: String) = map[key] as? String
            fun float(key: String) = (map[key] as? Number)?.toFloat()
            return AndroidOverlayConfiguration(
                title = string("title"),
                message = string("message"),
                backgroundColorHex = string("backgroundColorHex"),
                titleColorHex = string("titleColorHex"),
                messageColorHex = string("messageColorHex"),
                primaryButtonLabel = string("primaryButtonLabel"),
                primaryButtonBackgroundColorHex = string("primaryButtonBackgroundColorHex"),
                primaryButtonTextColorHex = string("primaryButtonTextColorHex"),
                secondaryButtonLabel = string("secondaryButtonLabel"),
                secondaryButtonBackgroundColorHex = string("secondaryButtonBackgroundColorHex"),
                secondaryButtonTextColorHex = string("secondaryButtonTextColorHex"),
                titleFontSize = float("titleFontSize"),
                messageFontSize = float("messageFontSize"),
                cornerRadius = float("cornerRadius"),
            )
        }
    }
}

File: lib/android_overlay_configuration.dart
/// Android-only styling for the full-screen overlay ("shield") that is drawn on
/// top of a blocked app.
///
/// Every colour is a hex string accepting `#RGB`, `#RRGGBB` or `#RRGGBBAA`
/// (alpha last). Any field left `null` keeps its built-in default, so you can
/// override just the pieces you care about.
///
/// This has no effect on iOS — use `ShieldConfigurationPayload` with
/// `FlutterScreenTime.setShieldConfiguration` for the iOS shield.
class AndroidOverlayConfiguration {
  /// Bold heading shown at the top of the overlay.
  final String title;

  /// Supporting line beneath the title. Hidden when null/empty.
  final String? message;

  /// Overlay background fill. Defaults to `#111827`.
  final String? backgroundColorHex;

  /// Colour of the [title] text. Defaults to `#FFFFFF`.
  final String? titleColorHex;

  /// Colour of the [message] text. Defaults to `#D1D5DB`.
  final String? messageColorHex;

  /// Label for the primary (dismiss) button. Defaults to `OK`.
  final String? primaryButtonLabel;

  /// Fill colour of the primary button. Defaults to `#2563EB`.
  final String? primaryButtonBackgroundColorHex;

  /// Text colour of the primary button. Defaults to `#FFFFFF`.
  final String? primaryButtonTextColorHex;

  /// Label for the optional secondary button. Hidden when null/empty.
  final String? secondaryButtonLabel;

  /// Fill colour of the secondary button. When null the button is outlined
  /// (transparent fill with a border in the text colour).
  final String? secondaryButtonBackgroundColorHex;

  /// Text colour of the secondary button. Defaults to `#FFFFFF`.
  final String? secondaryButtonTextColorHex;

  /// Title text size in sp. Defaults to `24`.
  final double? titleFontSize;

  /// Message text size in sp. Defaults to `16`.
  final double? messageFontSize;

  /// Corner radius (dp) applied to the buttons. Defaults to `12`.
  final double? cornerRadius;

  const AndroidOverlayConfiguration({
    required this.title,
    this.message,
    this.backgroundColorHex,
    this.titleColorHex,
    this.messageColorHex,
    this.primaryButtonLabel,
    this.primaryButtonBackgroundColorHex,
    this.primaryButtonTextColorHex,
    this.secondaryButtonLabel,
    this.secondaryButtonBackgroundColorHex,
    this.secondaryButtonTextColorHex,
    this.titleFontSize,
    this.messageFontSize,
    this.cornerRadius,
  });

  Map<String, dynamic> toMap() {
    return {
      "title": title,
      "message": message,
      "backgroundColorHex": backgroundColorHex,
      "titleColorHex": titleColorHex,
      "messageColorHex": messageColorHex,
      "primaryButtonLabel": primaryButtonLabel,
      "primaryButtonBackgroundColorHex": primaryButtonBackgroundColorHex,
      "primaryButtonTextColorHex": primaryButtonTextColorHex,
      "secondaryButtonLabel": secondaryButtonLabel,
      "secondaryButtonBackgroundColorHex": secondaryButtonBackgroundColorHex,
      "secondaryButtonTextColorHex": secondaryButtonTextColorHex,
      "titleFontSize": titleFontSize,
      "messageFontSize": messageFontSize,
      "cornerRadius": cornerRadius,
    };
  }
}
